src/agentscope/rag/knowledge_bank.py

```diff
@@ -0,0 +1,179 @@
+                    emb_model_name="qwen_emb_config",
+                    data_dirs_and_types={
+                        "../../docs/sphinx_doc/en/source/tutorial": [".md"],
+                    },
+                    persist_dir="./rag_storage/tutorial_assist",
+                )
+            ''
+        """
+        if knowledge_id in self.stored_knowledge:
+            raise ValueError(f"knowledge_id {knowledge_id} already exists.")
+
+        assert data_dirs_and_types is not None or knowledge_config is not None
+
+        if knowledge_config is None:
+            knowledge_config = copy.deepcopy(DEFAULT_INDEX_CONFIG)
+            for data_dir, types in data_dirs_and_types.items():
+                loader_config = copy.deepcopy(DEFAULT_LOADER_CONFIG)
+                loader_init = copy.deepcopy(DEFAULT_INIT_CONFIG)
+                loader_init["input_dir"] = data_dir
+                loader_init["required_exts"] = types
+                loader_config["load_data"]["loader"]["init_args"] = loader_init
+                knowledge_config["data_processing"].append(loader_config)
+
+        self.stored_knowledge[knowledge_id] = LlamaIndexKnowledge(
+            knowledge_id=knowledge_id,
+            emb_model=load_model_by_config_name(emb_model_name),
+            knowledge_config=knowledge_config,
+            model=load_model_by_config_name(model_name)
+            if model_name
+            else None,
+        )
+        logger.info(f"data loaded for knowledge_id = {knowledge_id}.")
+
+    def get_knowledge(
+        self,
+        knowledge_id: str,
+        duplicate: bool = False,
+    ) -> LlamaIndexKnowledge:
+        """
+        Get a Knowledge object from the knowledge bank.
+        Args:
+            knowledge_id (str):
+                unique id for the Knowledge object
+            duplicate (bool):
+                whether return a copy of the Knowledge object.
+        Returns:
+            LlamaIndexKnowledge:
+                the Knowledge object defined with Llama-index
+        """
+        if knowledge_id not in self.stored_knowledge:
+            raise ValueError(
+                f"{knowledge_id} does not exist in the knowledge bank.",
+            )
+        knowledge = self.stored_knowledge[knowledge_id]
+        if duplicate:
+            knowledge = copy.deepcopy(knowledge)
+        logger.info(f"knowledge bank loaded: {knowledge_id}.")
+        return knowledge
+
+    def equip(self, agent: AgentBase, duplicate: bool = False) -> None:
+        """
+        Equip the agent with the knowledge it requests
+        Args:
+            agent (AgentBase): the agent to be equipped with knowledge
+                if it has "rag_config"
+            duplicate (bool): whether to deepcopy the knowledge object
+        """
+        if hasattr(agent, "rag_config") and "knowledge_id" in agent.rag_config:
+            if not hasattr(agent, "knowledge_list"):
+                agent.knowledge_list = []
```

Comment:
this method is not compatible with rpc agent.  RpcAgent only has reply and observe methods inherited from `AgentBase`, we need to add a function to `AgentBase` to equip agent with the knowledge.

Reply:
added TODO to document RpcAgent support as future work